File: Assembly-FSharp/Assembly-FSharp/Managers/PuckManager.fs
namespace Styx.Managers

open UnityEngine
open Styx.Entities.PuckModule

type PuckManager () =
    inherit MonoBehaviour()
    
    let mutable puckState: PuckState = Unchecked.defaultof<PuckState>

    //*********************** SINGLETON ***************************************
    static let mutable instance: PuckManager = Unchecked.defaultof<PuckManager>
    static member Instance
        with get() = instance
    //*************************************************************************

    member this.Awake () = 
        if instance <> Unchecked.defaultof<PuckManager> && instance <> this then
            Object.Destroy (this.gameObject)
        else
            instance <- this

    member this.Register (newPuck: PuckState) =
        puckState <- newPuck
        puckState

    member this.GetPuck () =
        puckState
        
    member this.UpdatePuck (updatedPuck: Puck) =
        puckState <- { puckState with puck = updatedPuck }

    member this.SetVelocity (velocity: Vector3) =
        puckState.puck.rigidbodyModel.Rigidbody.velocity <- velocity
        puckState

File: Assets/_Scripts/Managers/PuckManager.cs
using Styx.Entities.PuckModule;
using UnityEngine;

namespace Styx.Managers
{
    public class PuckManager : MonoBehaviour
    {
        private PuckState _puckState;
        public static PuckManager Instance { get; private set; }

        public void Awake()
        {
            if (Instance != null && Instance != this)
                Destroy(gameObject);
            else
                Instance = this;

        }

        public void Register (PuckState puckState)
        {
            _puckState = puckState;
        }

        public PuckState GetPuckState()
        {
            return _puckState;
        }

        public void UpdatePuckState (PuckState updatedPuckState)
        {
            _puckState = updatedPuckState;
        }

        public void UpdateVelocity (Vector3 velocity)
        {
            _puckState.Puck.RigidbodyModel.Rigidbody.velocity = velocity;
        }

    }
}
